src/mo_frontend/bi_match.ml

```diff
@@ -384,12 +387,13 @@ let is_closed ctx t = if is_ctx_empty ctx then true else
   ConSet.disjoint ctx.var_set all_cons
 
 (** Raises when [er] is non-empty, optionally with a suggested type instantiation. *)
-let maybe_raise_underconstrained ctx env er =
+let maybe_raise_underconstrained ctx current_env unsolved er =
   let error_msg = ErrorUnderconstrained.to_string er in
   if error_msg = "" then () else
   let error_msg, hint =
-    let ts = List.map (fun c -> ConEnv.find c env) ctx.var_list in
-    if List.for_all (is_closed ctx) ts then
+    if ConSet.is_empty unsolved then
+      (* Future work: fill the unsolved and solved from previous rounds with holes, e.g. <_, Nat, _> *)
```

Comment:
https://github.com/dfinity/motoko/pull/5575 might be useful for 
1. Skipping solved tvars in the previous round (not necessary to provide)
2. Skipping unsolved tvars, so that we can always provide the hint, but then such a hint might not provide enough type info, we could continue solving...